### PR TITLE
Add per namespace pages that show all resource stats for a namespace

### PR DESCRIPTION
**Pull request by rmars**
A directory of namespaces can now be found on the Service Mesh page,
replacing the list of Deployments.

This PR moves the ServiceMesh page off of the `ListPods` endpoint entirely. This endpoint can now be removed. Also, moving the ServiceMesh page off of listing specific resources (other than namespaces) will allow #795 to be marked as complete.

This PR does the following:
- Replace the deployments table in the service mesh page with namespaces
- Add a Namespaces index page that lists all namespaces and their stats
- Add an individual namespace page showing all resources for that namespace
- Make the incomplete mesh message more generic to any resource type
- Revamp rest of service mesh page to move off ListPods

screen shot 2018-05-02 at 12 43 54 pm

screen shot 2018-05-01 at 3 15 23 pm
screen shot 2018-05-01 at 2 56 36 pm

### CODE REVIEW

File: web/app/js/components/Namespaces.jsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  loadFromServer() {
+    if (this.state.pendingRequests) {
+      return; // don't make more requests if the ones we sent haven't completed
+    }
+    this.setState({ pendingRequests: true });
+
+    this.api.setCurrentRequests(
+      _.map(["deployment", "replication_controller", "pod"], resource =>
```
Comment (siggy) [inline web/app/js/components/Namespaces.jsx]: not for this PR, but eventually we could consider modifying the `stat` endpoint to accept multiple resource types.

File: web/app/js/components/Namespaces.jsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    this.api = this.props.api;
+    this.handleApiError = this.handleApiError.bind(this);
+    this.loadFromServer = this.loadFromServer.bind(this);
+    this.state = this.getInitialState(this.props.location);
+  }
+
+  getInitialState(location) {
+    location = location || {};
+    let urlParams = new URLSearchParams(location.search);
+    let ns = urlParams.get("ns") || "default";
```
Comment (siggy) [inline web/app/js/components/Namespaces.jsx]: is there a way to force navigation to `/namespaces?ns=default`?
or, to put it another way, when i (manually) navigated to `/namespaces` without request params, i was expecting to see a list of namespaces, with links to each namespace page. related: should that namespaces page exist? just another type of `ResourceList`?

File: web/app/js/components/Namespaces.jsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        <MetricsTable
+          resource={friendlyTitle}
+          metrics={metrics}
+          api={this.api} />
+      </div>
+    );
+  }
+
+  render() {
+    let noMetrics = _.isEmpty(this.state.metrics.deploy) && _.isEmpty(this.state.metrics.rc);
```
Comment (siggy) [inline web/app/js/components/Namespaces.jsx]: check for `_.isEmpty(this.state.metrics.pod)`? i'm assuming there could be pods launched by things other than deployments and RCs, or, even by themselves?

File: web/app/js/components/ServiceMesh.jsx [hunk 4]
```diff
@@ -27,6 +25,58 @@ const serviceMeshDetailsColumns = [
...
+    sorter: (a, b) => numericSort(a.totalPods, b.totalPods),
+  },
+  {
+    title: "Mesh completion",
+    key: "meshification",
+    sorter: (a, b) => numericSort(a.meshedPercent.get(), b.meshedPercent.get()),
+    render: row => {
+      let containerWidth = 132;
+      let percent = row.meshedPercent.get();
+      let barWidth = percent < 0 ? 0 : Math.round(percent * containerWidth);
```
Comment (siggy) [inline web/app/js/components/ServiceMesh.jsx]: how does `percent < 0` happen?

File: web/app/js/components/ServiceMesh.jsx [hunk 5]
```diff
@@ -27,6 +25,58 @@ const serviceMeshDetailsColumns = [
     className: "numeric"
   }
 ];
+
+const barColor = percentMeshed => {
+  if (percentMeshed <= 0) {
```
Comment (siggy) [inline web/app/js/components/ServiceMesh.jsx]: related to other percent comment, how does `percentMeshed <= 0` happen? there's code further down that does `row.meshedPercent.get() > 0.9 ? "good" : "neutral";`. how does this relate to that?

File: web/app/js/components/Namespaces.jsx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+      _.map(["deployment", "replication_controller", "pod"], resource =>
+        this.api.fetchMetrics(this.api.urlsForResource[resource].url(this.state.ns).rollup))
+    );
+
+    Promise.all(this.api.getCurrentPromises())
+      .then(([deployRollup, rcRollup, podRollup]) => {
+        let includeConduitStats = this.state.ns === this.props.controllerNamespace; // allow us to get stats on the conduit ns
+        let deploys = processRollupMetrics(deployRollup, this.props.controllerNamespace, includeConduitStats);
+        let rcs = processRollupMetrics(rcRollup, this.props.controllerNamespace, includeConduitStats);
+        let pods = processRollupMetrics(podRollup, this.props.controllerNamespace, includeConduitStats);
```
Comment (siggy) [inline web/app/js/components/Namespaces.jsx]: i'm totally onboard with displaying info about the conduit namespace. i'm realizing we should do the same in Grafana, which currently disallows displaying of controller deployment info. i've filed #908 to track this.

File: web/app/js/components/Namespaces.jsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+import ConduitSpinner from "./ConduitSpinner.jsx";
+import ErrorBanner from './ErrorBanner.jsx';
+import MetricsTable from './MetricsTable.jsx';
+import PageHeader from './PageHeader.jsx';
+import { processRollupMetrics } from './util/MetricUtils.js';
+import React from 'react';
+import './../../css/list.css';
+import 'whatwg-fetch';
+
+export default class Namespaces extends React.Component {
```
Comment (siggy) [inline web/app/js/components/Namespaces.jsx]: something to consider down the road, the namespace-specific pages have no reference/context in the navigation. consider a tree or some other way to indicate to the user where they are, and how they could navigate to/from that page.

File: web/app/js/components/Namespaces.jsx (hunk 1 above)
Comment (rmars) [reply web/app/js/components/Namespaces.jsx]: Filed #911

File: web/app/js/components/Namespaces.jsx (hunk 2 above)
Comment (rmars) [reply web/app/js/components/Namespaces.jsx]: Yep you're right; at first I thought that ServiceMesh would be the only index into namespaces, but it's quite easy to just add another page that lists stats by Namespace (just another ResourceList, as you mentioned!). So I've added that.

File: web/app/js/components/ServiceMesh.jsx (hunk 4 above)
Comment (rmars) [reply web/app/js/components/ServiceMesh.jsx]: Percentage.js returns -1 if the denominator is 0:

```
export default function Percentage(numerator, denominator) {
  this.decimal = -1;
  if (denominator !== 0) {
    this.decimal = numerator / denominator;
  }
}
```

File: web/app/js/components/Namespaces.jsx (hunk 6 above)
Comment (rmars) [reply web/app/js/components/Namespaces.jsx]: thanks!!

File: web/app/js/components/Namespaces.jsx (hunk 7 above)
Comment (rmars) [reply web/app/js/components/Namespaces.jsx]: Yeah! I've added a namespace index page per your other comment; so now you can navigate to individual namespace pages via the ServiceMesh page or via the Namespace index page at `/namespaces`

File: web/app/js/components/ServiceMesh.jsx (hunk 5 above)
Comment (rmars) [reply web/app/js/components/ServiceMesh.jsx]: A percent meshed < 0 can happen if there are 0 total pods (denominator is 0, so the percentage helper returns -1 as the decimal and "N/A" as the pretty formatted).

The barColor is independent of this; it's just deciding whether to color the bar green or grey (and in the future or red) based on the percentage of pods that have already been meshed.

File: web/app/js/components/Namespaces.jsx (hunk 2 above)
Comment (rmars) [reply web/app/js/components/Namespaces.jsx]: screen shot 2018-05-08 at 12 40 03 pm

**Pull request merged**